### ISSUE TITLE
Documenting strategies for embedding Dash apps in an existing Flask app

### ISSUE THREAD

**Issue by ned2** [plotly/dash-docs#246]
The community has been asking for a while about strategies for adding a Dash app to an existing Flask app (see plotly/dash#214). I think it would be good if we could get something into the documentation to help with this.

I'd be up for putting together a PR, but first I was hoping to get @plotly/dash's thoughts on the different strategies we could endorse, as well as some ideas about when the different approaches might be useful.

I'm currently aware of two distinct strategies:

1. Passing your `Dash` instance constructor the `server` kwarg that is an existing `Flask` instance as well as passing a value to the `url_base_pathname` kwarg so that multiple Dash apps can sit side by side without clashing. This is described very well in [a comment on the above mentioned issue](https://github.com/plotly/dash/issues/214#issuecomment-434655642).

2. Using Werkzeug's `DispatcherMiddleware` to mount the `Flask` instance of one or more existing `Dash` instances at specific routes of a parent `Flask` instance.  This is discussed in this [Dash community forum thread](https://community.plot.ly/t/hosting-multiple-dash-apps-with-uwsgi-nginx/6758).

My take is that the first option has less moving parts and so is a bit simpler, however the second has a bit more flexibility in that you can embed arbitrary existing Dash apps that can also be run independently (as opposed to requiring an existing flask instance to be passed in). 

Anyone have more strategies to add or thoughts about these two?

**Comment by emenchini** [comment]
Bump. It would be great to have an developer-official best practice in the docs.

**Comment by sidd-hart** [comment]
The first option isn't very useful if you're trying to add a flask based login (authentication) page to the dash app.

**Comment by okomarov** [comment]
@ned2 I created a solution in https://github.com/okomarov/dash_on_flask for embedding a Dash application into a Flask app which:
* uses the application factory pattern
* uses the Flask-Login (and other extensions)

Explanations at:
https://medium.com/@olegkomarov_77860/how-to-embed-a-dash-app-into-an-existing-flask-app-ea05d7a2210b

Let me know if something is re-usable for your purposes.

**Comment by berryt08** [comment]
@ned2, great idea I have successfully integrated Dash into a Flask app with multiple HTML pages as well as flask-login on the Dash app, however it took me quite some time to figure out and then clean up.

@sidd-hart, as mentioned above this can be done. I used flask-login and wtforms for my authentication and protected my dash app like so:
`def protect_views(app):
    for view_func in app.server.view_functions:
        if view_func.startswith(app.url_base_pathname):
            app.server.view_functions[view_func] = login_required(app.server.view_functions[view_func])
    return app

server = Flask(__name__)
app = Dash(__name__, server=server, url_base_pathname='/app')

app = protect_views(app)`

Obviously there is lots of Flask-Login code in the configuration as well but I'm sure you are aware of that already.

Hope this helps.
